proposals/0046-actions.md

```diff
@@ -0,0 +1,637 @@
+
+Actions work well when client JavaScript is enabled. However, you may see unexpected behavior when a form is submitted _before_ your component's JavaScript loads. This is common on slow internet connections or older devices. To offer a fallback experience, you can add a progressive fallback to your form.
+
+To add a fallback, add the `method="POST"` property to your `<form>` element. Then, call the `getNameProps()` function from `astro:actions` with the action you want to use (ex. `getNameProps(actions.comment)`). Spread the result onto an `input` within your form to apply metadata for Astro to handle the request:
+
+```tsx
+import { actions, getNameProps } from 'astro:actions';
+
+// src/components/Comment.jsx
+export function Comment({ postId }: { postId: string }) {
+	return (
+		<form method="POST" onSubmit={...}>
+			<input {...getNameProps(actions.comment)} />
+			{/* result:
+			<input
+				type="hidden"
+				name="_astroAction"
+				value="/_actions/comment" />
+			*/}
+		</form>
+	)
+}
+```
+
+Implementation: The `getNameProps()` function returns hidden input attributes to tell Astro which action should handle the request. Using middleware, Astro will intercept incoming requests with a form data `Content-type`. The form data is parsed to check for the `_astroAction` field and call the appropriate action.
+
+### Handle an action result on the server
+
+When using a progressive fallback, you can get the result of an action from your Astro frontmatter. Call `Astro.getActionResult()` with the action you want (ex. `Astro.getActionResult(actions.comment)`). This will return type-safe data when a matching POST request is received and `undefined` otherwise.
+
+```astro
+---
+import { actions } from 'astro:actions';
+
+const comment = Astro.getActionResult(actions.comment);
+---
+
+{comment && (
+	<article class="new-comment">
+		{/* ... */}
+	</article>
+)}
+```
+
+## Handle errors
+
+Actions can raise exceptions for any number of reasons: an invalid input, an unauthorized request, a missing resource, etc. By default, these errors will `throw` wherever your action is called.
+
+To handle errors in your code without using a `try / catch`, you can chain the `.safe()` function onto your existing actions. This will change the return type to either a `data` object with the result, or an `error` object with the exception.
+
+You can call `.safe()` from a client component to handle `data` and `error`:
+
+```tsx
+// src/components/Like.tsx
+import { actions } from 'astro:actions';
+
+export function Like() {
+	return (
+		<button onClick={async () => {
+			const { data, error } = await actions.like.safe({ postId });
+			if (data) // set state
+			// otherwise, handle the error
+		}}>
+			{likes} likes
+		</button>
+	)487918
+}
+```
+
+You can also use `.safe` when getting the action result from your Astro frontmatter. Add `.safe` to the end of your action when passing to `Astro.getActionResult()`:
```

Comment:
Echoing my above comments on `safe()`, it feels even more necessary here. `getActionResult` is a way to prevent needing to check the method, etc. It allows you to keep your code flat and only operate if it returns a result. So having a throwing-form makes the whole thing way less nice to me.

Reply:
So given the above conservation, does `Astro.getActionResult()` currently throw? It doesn't say that it does.

Reply:
Yes, `getActionResult()` would throw the same way as the client. I agree `.safe` would be a more preferred pattern on the server, since you're more likely wanting to respect validation errors on the server. Still, wanted to keep the two APIs consistent. Do you have thoughts?

Reply:
I don't see them as being the same API, so I don't think consistency doesn't feel as important. In the case of the RPC you are:

1. Calling an action.
2. Getting the result of that action.

And `getActionResult` is only (2). It's even in the name `getActionResult`, *getting* something shouldn't throw, imo. The result is either data or an error, and you are just requesting that info.

Reply:
This is a good point. The real "smell" for me was the inconsistent throwing behavior. The action would not throw _unless_ you call `getActionResult()`. If throwing is really what we want, it should throw on the backend when it's unhandled as well.

I agree the name should imply it's safe only. I'm on board with flipping that behavior